lib/Dialect/Torch/Transforms/ScalarizeShapes.cpp

```diff
@@ -0,0 +1,341 @@
+
+    Value cstNone = rewriter.create<Torch::ConstantNoneOp>(loc);
+    Value cstFalse = rewriter.create<Torch::ConstantBoolOp>(
+        loc, rewriter.getBoolAttr(false));
+    rewriter.replaceOpWithNewOp<Torch::AtenTensorOp>(
+        op, op.getType(), dimList, cstNone, cstNone, cstFalse);
+    return success();
+  }
+};
+} // namespace
+
+namespace {
+class PropagateAtenIndexSelectPattern
+    : public OpRewritePattern<AtenIndexSelectOp> {
+public:
+  using OpRewritePattern<AtenIndexSelectOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(AtenIndexSelectOp op,
+                                PatternRewriter &rewriter) const override {
+    auto loc = op.getLoc();
+
+    SmallVector<Value> elements;
+    if (failed(getListFromTensor(op.getSelf(), elements)))
+      return failure();
+
+    int64_t dim;
+    if (!matchPattern(op.getDim(), m_TorchConstantInt(&dim)))
+      return rewriter.notifyMatchFailure(op, "requires a constant dim");
+
+    DenseElementsAttr idx;
+    if (!matchPattern(op.getIndex(), m_Constant(&idx)))
+      return rewriter.notifyMatchFailure(op, "requires a constant index");
+
+    auto selfTy = cast<BaseTensorType>(op.getSelf().getType());
+    if (!selfTy.hasSizes())
+      return rewriter.notifyMatchFailure(op, "requires known rank");
+
+    auto selfShape = selfTy.getSizes();
+    int64_t selfRank = selfShape.size();
+    dim = dim < 0 ? dim + selfRank : dim;
+    int64_t dimLength = elements.size();
+    if (selfShape[dim] != dimLength)
+      return rewriter.notifyMatchFailure(
+          op, "dim length does not match number of elements");
+
+    for (int64_t i = 0; i < selfRank; ++i) {
+      if (i == dim)
+        continue;
+      if (selfShape[i] != 1)
+        return rewriter.notifyMatchFailure(op,
+                                           "expects unary non-dim dimension");
+    }
+
+    SmallVector<Value> selected;
+    if (idx.isSplat()) {
+      int64_t indexInt = idx.getSplatValue<APInt>().getSExtValue();
+      indexInt = indexInt < 0 ? indexInt + dimLength : indexInt;
+      selected.resize(idx.getNumElements(), elements[indexInt]);
+    } else {
+      for (APInt val : idx.getValues<APInt>()) {
+        int64_t indexInt = val.getSExtValue();
+        selected.push_back(elements[indexInt]);
+      }
+    }
+
+    auto eTy = elements.front().getType();
+
+    auto dimList = rewriter.create<Torch::PrimListConstructOp>(
+        loc, rewriter.getType<Torch::ListType>(eTy), selected);
+
+    Value cstNone = rewriter.create<Torch::ConstantNoneOp>(loc);
+    Value cstFalse = rewriter.create<Torch::ConstantBoolOp>(
+        loc, rewriter.getBoolAttr(false));
+    rewriter.replaceOpWithNewOp<Torch::AtenTensorOp>(
+        op, op.getType(), dimList, cstNone, cstNone, cstFalse);
+    return success();
+  }
+};
+} // namespace
+
+namespace {
+class PropagateAtenSliceTensorPattern
+    : public OpRewritePattern<AtenSliceTensorOp> {
+public:
+  using OpRewritePattern<AtenSliceTensorOp>::OpRewritePattern;
```

Comment:
this could use a few words of comment explaining when it applies. From my reading it seems to do this:
- find aten.slice.tensor(aten.tensor(x), dims)
- return aten.tensor(x[dims])
- where x is a list of values
- and the tensor has to have shape 1...1, x, 1...1

Reply:
Added